File: 1155. Number of Dice Rolls With Target Sum/1155. Number of Dice Rolls With Target Sum bottom-up.java
class Solution {
    public int numRollsToTarget(int n, int k, int target) {
        int[][] dp = new int[n + 1][target + 1];
        int modulo = 1000000007;
        dp[0][0] = 1;
        for (int dice_left = 1; dice_left <= n; dice_left++) {
            for (int target_left = 1; target_left <= target; target_left++) {
                for (int i = 1; i <= Math.min(target_left, k); i++) {
                    dp[dice_left][target_left] += dp[dice_left - 1][target_left - i];
                    dp[dice_left][target_left] %= modulo;
                }
            }
        }
        return dp[n][target];
    }
}

File: 1155. Number of Dice Rolls With Target Sum/1155. Number of Dice Rolls With Target Sum bottom-up.py
from functools import lru_cache


class Solution:
    def numRollsToTarget(self, n: int, k: int, target: int) -> int:
        modulo = 10**9+7
        dp = [[0]*(target+1) for _ in range(n+1)]
        dp[0][0] = 1

        for dice_left in range(1, n+1):
            for target_left in range(1, target+1):
                for i in range(1, min(k, target_left)+1):
                    dp[dice_left][target_left] += dp[dice_left-1][target_left-i]
                dp[dice_left][target_left] %= modulo
        return dp[n][target]


n = 30
k = 30
target = 500
s = Solution()
print(s.numRollsToTarget(n, k, target))
